plugins/bcm57xx/fu-bcm57xx-device.c

```diff
@@ -0,0 +1,860 @@
+	img_ape = fu_firmware_get_image_by_id (firmware_tmp, "ape", NULL);
+	if (img_ape != NULL)
+		fu_firmware_add_image (firmware, img_ape);
+
+	/* success */
+	return g_steal_pointer (&firmware);
+}
+
+static gboolean
+fu_bcm57xx_device_write_firmware (FuDevice *device,
+				  FuFirmware *firmware,
+				  FwupdInstallFlags flags,
+				  GError **error)
+{
+	FuBcm57xxDevice *self= FU_BCM57XX_DEVICE (device);
+	const guint8 *buf;
+	gsize bufsz = 0;
+	gsize bufsz_dwrds = fu_device_get_firmware_size_max (FU_DEVICE (self)) / sizeof(guint32);
+	g_autofree guint32 *buf_dwrds = g_new0 (guint32, bufsz_dwrds);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autoptr(FuDeviceLocker) locker2 = NULL;
+	g_autoptr(GBytes) blob = NULL;
+
+	/* build the images into one linear blob of the correct size */
+	fu_device_set_status (device, FWUPD_STATUS_DECOMPRESSING);
+	blob = fu_firmware_write (firmware, error);
+	if (blob == NULL)
+		return FALSE;
+
+	/* align into uint32_t buffer */
+	buf = g_bytes_get_data (blob, &bufsz);
+	for (guint i = 0; i < bufsz_dwrds; i++) {
+		if (!fu_memcpy_safe ((guint8 *) buf_dwrds,
+				     bufsz_dwrds * sizeof(guint32),
+				     i * sizeof(guint32),		/* dst */
+				     buf, bufsz, i * sizeof(guint32),	/* src */
+				     sizeof(guint32), error))
+			return FALSE;
+	}
+
+	/* hit hardware */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	locker = fu_device_locker_new_full (self,
+					    (FuDeviceLockerFunc) fu_bcm57xx_device_nvram_acquire_lock,
+					    (FuDeviceLockerFunc) fu_bcm57xx_device_nvram_release_lock,
+					    error);
+	if (locker == NULL)
+		return FALSE;
+	locker2 = fu_device_locker_new_full (self,
+					     (FuDeviceLockerFunc) fu_bcm57xx_device_nvram_enable_write,
+					     (FuDeviceLockerFunc) fu_bcm57xx_device_nvram_disable,
+					     error);
+	if (locker2 == NULL)
+		return FALSE;
+	if (!fu_bcm57xx_device_nvram_write (self, 0x0, buf_dwrds, bufsz_dwrds, error))
+		return FALSE;
+
+	/* reset APE */
+	if (!fu_bcm57xx_device_reset_ape (self, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_bcm57xx_device_setup (FuDevice *device, GError **error)
+{
+	FuBcm57xxDevice *self = FU_BCM57XX_DEVICE (device);
+	guint32 fwversion = 0;
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	fu_bcm57xx_device_ensure_ethtool_iface (self);
+	locker = fu_device_locker_new_full (self,
+					    (FuDeviceLockerFunc) fu_bcm57xx_device_nvram_enable,
+					    (FuDeviceLockerFunc) fu_bcm57xx_device_nvram_disable,
+					    error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* get NVRAM version */
+	if (!fu_bcm57xx_device_nvram_read (self, BCM_NVRAM_STAGE1_BASE + BCM_NVRAM_STAGE1_VERSION,
+					   &fwversion, 1, error))
+		return FALSE;
+	if (fwversion != 0x0) {
+		g_autofree gchar *fwversion_str = NULL;
+
+		/* this is always 0x0 on the non-free firmware */
+		fu_device_set_version_format (device, FWUPD_VERSION_FORMAT_TRIPLET);
+		fu_device_set_version (device, fwversion_str);
```

Comment:
This is NULL here and causes a failure when trying to flash.

 `Device NetXtreme BCM5719 Gigabit Ethernet PCIe [48e3a2f898a9e9d83dd82699ca2c453b5f52fc36] has no firmware version`

Reply:
I had the commands the wrong way around, fixed, thanks.